### PR TITLE
send resource count to telemetry

### PR DESCRIPTION
**Pull request by tanay-microsoft**
## Description
Send resource count to telemetry:
## Org Resources
- [*] installed extensions
## Project Resources
- [*] repositories
- [*] testplans
- [*] builds
- [*] releases
- [*] taskgroups

### CODE REVIEW

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 1]
```diff
@@ -43,11 +43,20 @@ class SVTResourceResolver: AzSKRoot {
     hidden [string[]] $VariableGroupIds = @();
     hidden [bool] $isServiceIdBasedScan = $false;
 
+    $orgTelemetryData = @{
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:46]: remove this

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 2]
```diff
@@ -476,6 +565,9 @@ class SVTResourceResolver: AzSKRoot {
                                         $taskAgentQueues = $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues | Where-Object {($_.pool.isLegacy -eq $false) -and ($this.AgentPools -contains $_.name) } 
                                     }
                                 }
+                                if ($null -ne $taskAgentQueues) {
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:568]: what if a specific name of agent pool was specified in scan command? - that will make length of $taskAgentQueues equal to 1.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 3]
```diff
@@ -527,6 +619,10 @@ class SVTResourceResolver: AzSKRoot {
                                 }
                             }
 
+                            if($null -ne $varGroups) {
+                                $projectData["variablegroups"] = $varGroups.Length;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:623]: what if a specific name of var grp was specified in scan command? - that will make length of $varGrps equal to 1.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 4]
```diff
@@ -294,6 +379,8 @@ class SVTResourceResolver: AzSKRoot {
                                    
                                     if (--$nObj -eq 0) { break; } 
                                 }
+                                # builds count here
+                                $projectData['build'] = $buildDefnsObj.Length;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:383]: try using Measure-object just to stay consistent with src code.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 5]
```diff
@@ -463,6 +557,10 @@ class SVTResourceResolver: AzSKRoot {
                             if (([Helpers]::CheckMember($agentPoolsDefnsObj, "fps.dataProviders.data") ) -and (($agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider") -and $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues)) {
                                 $nObj = $this.MaxObjectsToScan
                                 $taskAgentQueues = $null;
+                                if ($null -ne $agentPoolsDefnsObj) {
+                                    $allAgentPools = $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues | where-object{$_.pool.isLegacy -eq $false};
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:561]: remove the isLegacy flag.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 6]
```diff
@@ -463,6 +557,10 @@ class SVTResourceResolver: AzSKRoot {
...
+                                    $projectData["agentpools"] = $allAgentPools.Length
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:562]: Use ($x | Measure-Object).Count

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 7]
```diff
@@ -514,6 +611,9 @@ class SVTResourceResolver: AzSKRoot {
                         if (([Helpers]::CheckMember($variableGroupObj, "count") -and $variableGroupObj[0].count -gt 0) -or (($variableGroupObj | Measure-Object).Count -gt 0 -and [Helpers]::CheckMember($variableGroupObj[0], "name"))) {
                     
                             $varGroups = $null;
+                            if($null -ne $variableGroupObj) {
+                                $projectData["variablegroups"] = $variableGroupObj.Length;
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:615]: Use ($x | Measure-Object).Count

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 8]
```diff
@@ -245,6 +321,16 @@ class SVTResourceResolver: AzSKRoot {
                 {
                     $projectName = $thisProj.name
                     $projectId = $thisProj.id;
+                    [Hashtable] $projectData = @{
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:324]: add for svc connection also.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 9]
```diff
@@ -157,6 +162,77 @@ class SVTResourceResolver: AzSKRoot {
         }
     }
 
+    [void] GetResourceCount($projectName, $organizationId, $projectId, $projectData, $topNQueryString) {
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:165]: why $topNQueryString ?

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 10]
```diff
@@ -157,6 +162,77 @@ class SVTResourceResolver: AzSKRoot {
...
+        # fetching the repository count of a project
+        try{
+            $resourceURL = "https://dev.azure.com/$($this.organizationName)/$($projectName)/_apis/git/repositories?api-version=6.0"
+            $responseList = [WebRequestHelper]::InvokeGetWebRequest($resourceURL) ;
+            # $this.AddSVTResource("Repos", $projectName, "ADO.Repo", "organization/$organizationId/project/$projectId", $null, "");
+            $projectData['repositories'] = $responseList.Length
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:171]: use Measure-Object

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 11]
```diff
@@ -157,6 +162,77 @@ class SVTResourceResolver: AzSKRoot {
...
+            $responseList = [WebRequestHelper]::InvokeGetWebRequest($resourceURL) ;
+            $projectData['taskgroups'] = $responseList.Length
+            # fetch the builds count if provided builds not equals to *
+            if($projectData['build'] -eq -1) {
+                $resourceURL = ""
+                if ([string]::IsNullOrEmpty($topNQueryString)) {
+                    $topNQueryString = '&$top=10000'
+                    $resourceURL = ("https://dev.azure.com/{0}/{1}/_apis/build/definitions?api-version=4.1&queryOrder=lastModifiedDescending" +$topNQueryString) -f $($this.SubscriptionContext.SubscriptionName), $projectName;
+                }
+                else {
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:187]: else condition is not required.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 12]
```diff
@@ -157,6 +162,77 @@ class SVTResourceResolver: AzSKRoot {
...
+            }
+            # fetch the agent pools count
+            if($projectData["agentpools"] -eq -1) {
+                $agentPoolsDefnURL = ("https://dev.azure.com/{0}/{1}/_settings/agentqueues?__rt=fps&__ver=2") -f $($this.SubscriptionContext.SubscriptionName), $projectName;
+                try {
+                    $agentPoolsDefnsObj = [WebRequestHelper]::InvokeGetWebRequest($agentPoolsDefnURL);
+                    if (([Helpers]::CheckMember($agentPoolsDefnsObj, "fps.dataProviders.data") ) -and (($agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider") -and $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues)) {
+                        $nObj = $this.MaxObjectsToScan
+                        $taskAgentQueues = $null;
+                        $taskAgentQueues = $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues | where-object{$_.pool.isLegacy -eq $false};
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:207]: isLegacy check is not required.
$maxObj is not reqd.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 13]
```diff
@@ -157,6 +162,77 @@ class SVTResourceResolver: AzSKRoot {
...
+                    }
+                }
+                catch {}
+            }
+        }
+        catch {}
+        # fetch the variable groups count
+        if ($projectData["variablegroups"] -eq -1) {
+            $variableGroupURL = ("https://dev.azure.com/{0}/{1}/_apis/distributedtask/variablegroups?api-version=6.1-preview.2") -f $($this.organizationName), $projectId;
+            try{
```
Comment (abhaydaga) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:220]: why is this added in a separate try-catch?

it is better then to segregate each resource type computation in individual try-catch blocks.

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 (hunk 9 above)
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:165]: Add this method in the last. let main logic of svt resolver come first

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 14]
```diff
@@ -157,6 +162,77 @@ class SVTResourceResolver: AzSKRoot {
...
+            }
+            # fetch the release count if provided builds not equals to *
+            if($projectData['release'] -eq -1) {
+                $resourceURL = ("https://vsrm.dev.azure.com/{0}/{1}/_apis/release/definitions?api-version=4.1-preview.3" +$topNQueryString) -f $($this.SubscriptionContext.SubscriptionName), $projectName;
+                $responseList = [WebRequestHelper]::InvokeGetWebRequest($resourceURL);
+                $projectData['release'] = $responseList.Length;
+            }
+            # fetch the agent pools count
+            if($projectData["agentpools"] -eq -1) {
+                $agentPoolsDefnURL = ("https://dev.azure.com/{0}/{1}/_settings/agentqueues?__rt=fps&__ver=2") -f $($this.SubscriptionContext.SubscriptionName), $projectName;
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:201]: Try to use existing code(even if small modification required), avoid writing same code in multiple places, we already have api calls for resources.

**Review by Garima-msft — DISMISSED** [review]

**Review by abhaydaga — APPROVED** [review]

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 15]
```diff
@@ -14,6 +14,8 @@ class SVTResourceResolver: AzSKRoot {
     [int] $MaxObjectsToScan;
     [SVTResource[]] $SVTResources = @();
     [int] $SVTResourcesFoundCount = 0;
+
+	[bool] $IsAIEnabled = $false;
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:18]: indentation

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 16]
```diff
@@ -48,6 +50,9 @@ class SVTResourceResolver: AzSKRoot {
         $this.MaxObjectsToScan = $MaxObj #default = 0 => scan all if "*" specified...
         $this.SetallTheParamValues($organizationName, $ProjectNames, $BuildNames, $ReleaseNames, $AgentPools, $ServiceConnectionNames, $VariableGroupNames, $ScanAllArtifacts, $PATToken, $ResourceTypeName, $AllowLongRunningScan, $ServiceId, $IncludeAdminControls);            
         $this.skipOrgUserControls = $skipOrgUserControls
+		if ([RemoteReportHelper]::IsAIOrgTelemetryEnabled()) { 
+			$this.IsAIEnabled = $true; 
+		}
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:55]: indentation

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 17]
```diff
@@ -692,4 +718,56 @@ class SVTResourceResolver: AzSKRoot {
         
         return [AdministratorHelper]::isUserProjectAdminMember($this.organizationName, $project, $allowedAdminGrp);
     }
+
+    # getting resources count and sending them to telemetry as well
+    [void] GetResourceCount($projectName, $organizationId, $projectId, $projectData) {
+        try{
+            # fetching the repository count of a project
+            $resourceURL = "https://dev.azure.com/$($this.organizationName)/$($projectName)/_apis/git/repositories?api-version=6.0"
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:726]: not needed double $

File: src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 18]
```diff
@@ -692,4 +718,56 @@ class SVTResourceResolver: AzSKRoot {
...
+            $responseList = [WebRequestHelper]::InvokeGetWebRequest($resourceURL);
+            $projectData['release'] = ($responseList | Measure-Object).Count;
+
+            # fetch the agent pools count
+            if($projectData["agentPools"] -eq -1) {
+                $agentPoolsDefnURL = ("https://dev.azure.com/{0}/{1}/_settings/agentqueues?__rt=fps&__ver=2") -f $($this.SubscriptionContext.SubscriptionName), $projectName;
+                $agentPoolsDefnsObj = [WebRequestHelper]::InvokeGetWebRequest($agentPoolsDefnURL);
+                if (([Helpers]::CheckMember($agentPoolsDefnsObj, "fps.dataProviders.data") ) -and (($agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider") -and $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues)) {
+                    $taskAgentQueues = $agentPoolsDefnsObj.fps.dataProviders.data."ms.vss-build-web.agent-queues-data-provider".taskAgentQueues;
+                    $projectData["agentPools"] = ($taskAgentQueues | Measure-Object).Count
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:756]: Service connection is not added? like other resources count added here

Comment (tanay-microsoft) [reply src/AzSK.ADO/Framework/Core/SVT/SVTResourceResolver.ps1:756]: It is being evaluated at line 430, so no need to add here
